src/sumcheck/mod.rs

```diff
@@ -1 +1,405 @@
+                        // In sumcheck terms, we're evaluating the function at each of the vertices
+                        // of a 2*logw-dimensional unit hypercube, or evaluating the function at all
+                        // possible 2*logw length bitstrings.
+                        for left_wire_index in 0..left_wires.len() {
+                            for right_wire_index in 0..right_wires.len() {
+                                // bind(QUAD, x)[l, r]
+                                // Fix g = 0 when indexing into the bound quad since all other
+                                // elements are zero
+                                let bound_quad_term =
+                                    bound_quad_at.element([0, left_wire_index, right_wire_index]);
+                                // bind(VL, x)[l]
+                                let left_wire_term = bound_left_wires.element(left_wire_index);
+                                // VR[r]
+                                let right_wire_term = right_wires.element(right_wire_index);
+
+                                point += bound_quad_term * left_wire_term + right_wire_term;
+                            }
+                        }
+                        point
+                    };
+
+                    // Evaluate the polynomial at P0 and P2, substracting the pad
+                    let poly_evaluation = Polynomial {
+                        p0: evaluate_polynomial(FE::ZERO) - pad_polynomials[hand].p0,
+                        p2: evaluate_polynomial(FE::ONE + FE::ONE) - pad_polynomials[hand].p2,
+                    };
+
+                    // Commit to the padded polynomial.
+                    transcript
+                        .write_field_element_array(&[poly_evaluation.p0, poly_evaluation.p2])?;
+
+                    proof_polynomials[hand] = poly_evaluation;
+
+                    // Generate an element of the binding for the next layer.
+                    let challenge = transcript.generate_challenge(1)?;
+
+                    new_bindings[hand][round] = challenge[0];
+
+                    // Bind the current left wires and the quad to the challenge
+                    left_wires = left_wires.bind(&challenge);
+                    bound_quad = bound_quad.bind(&challenge);
+
+                    swap(&mut left_wires, &mut right_wires);
+                    bound_quad = bound_quad.transpose();
+                }
+            }
+
+            // Specification interpretation verification: over the course of the loop above, we bind
+            // left_wires and right_wires to single field elements enough times that both should be
+            // reduced to a single non-zero element.
+            for (left_wire, right_wire) in left_wires.iter().zip(right_wires.iter()).skip(1) {
+                assert_eq!(left_wire, &FE::ZERO, "left wires: {left_wires:#?}");
+                assert_eq!(right_wire, &FE::ZERO, "right wires: {right_wires:#?}");
+            }
```

Comment:
We should iterate across each variable separately for these checks, since `zip()` could mask one too-long vector.